### ISSUE TITLE
Use completer in notebook example

### ISSUE THREAD

**Issue by mattma** [jupyterlab/jupyterlab#2172]
I tried out [Notebook example](https://github.com/jupyterlab/jupyterlab/blob/master/examples/notebook/src/index.ts) and I want to integrate with [jupyterlab/completer](https://github.com/jupyterlab/jupyterlab/tree/master/packages/completer) package to autocomplete commands like in the official [Application Example](https://github.com/jupyterlab/jupyterlab/tree/master/examples/app).

I noticed that it use [Register Completer Extension Plugin](https://github.com/jupyterlab/jupyterlab/blob/master/examples/app/index.js#L45) in the **Application Example**. By reading the [source code of Completer Extension](https://github.com/jupyterlab/jupyterlab/blob/master/packages/completer-extension/src/index.ts). I do not know how to add completer feature into [Notebook example](https://github.com/jupyterlab/jupyterlab/blob/master/examples/notebook/src/index.ts).

There is [No JupyterLab Application required](https://github.com/jupyterlab/jupyterlab/blob/master/examples/app/index.js#L4) in **Notebook Example** so that I cannot use [registerPluginModules method](https://github.com/jupyterlab/jupyterlab/blob/master/examples/app/index.js#L45) to add *completer*. It should be a simple way to integrate with **Notebook example**. Hope that someone can show me how? Thanks in advance!

**Comment by blink1073** [comment]
Hi @mattma, agreed that this should be part of the example.  I'm marking this as 1.0 since we are focusing on getting the beta out, but if you'd like to tackle it, it would involve using the completer extension as a guide but removing the completer manager.  You would reuse the code in the [notebook](https://github.com/jupyterlab/jupyterlab/blob/master/packages/completer-extension/src/index.ts#L180) portion, but instantiate the [handler](https://github.com/jupyterlab/jupyterlab/blob/master/packages/completer-extension/src/index.ts#L180) directly as in [here](https://github.com/jupyterlab/jupyterlab/blob/master/packages/completer-extension/src/index.ts#L91).  Instead of calling the `select` and `invoke` commands, you would select and invoke on your singleton completer handler.

**Comment by mattma** [comment]
@blink1073 Thanks for your suggestion. It works as expected. 

While i work on completer integration, I realized that you and contributor just rename **CompleterWidget** to **Completer** in the source code one day ago (*_-). Several things I learned along the way. One of the hardest lessons is about **Completer Widget isAttached property is not set by setFlag**. It took me a while to figure out why **update-request** message has never been triggered since **invoke-message** was successfully boardcasted.

I will try to spend some time on cleaning up the code. And contribute back to the **Notebook Example** with a PR. :)

**Comment by blink1073** [comment]
Thanks for the feedback, and looking forward to your PR!

**Comment by mattma** [comment]
Will probably submit this weekend. 

Another question, instead of hitting `tab` to trigger completer, can I hit the "dot" key, auto trigger it?

Example

```
import datetime
datetime.
```

By hitting the `datetime` then `.` will auto trigger completer with list of hints? 

Here is [command key binding](https://github.com/jupyterlab/jupyterlab/blob/master/examples/notebook/src/index.ts#L275), so I tried with `keys: ['.'],` with completer, that does not work.

**Comment by blink1073** [comment]
Hmm, I think that the command that handled a dot would have to add the dot to the string it is sending for the completion request, since it wouldn't yet be in the text buffer.

**Comment by mattma** [comment]
@blink1073 [PR](https://github.com/jupyterlab/jupyterlab/pull/2210) is raised. I have tested in my local environment and it works fine.

For the dot trigger autocomplete, I have tried below

```
    {
      selector: '.jp-Notebook.jp-mod-editMode .jp-mod-completer-enabled',
      keys: ['.'],
      command: cmdIds.invokeNotebook
    },
```

Issue above, when I hit `.` on my keyboard, it will autocomplete the existing command, but it won't insert the text `.` in the line. Ex: `datetime.` is what I want, instead of `datetime`, `.` won't be inserted.

```
    {
      selector: '.jp-Notebook.jp-mod-editMode .jp-mod-completer-enabled',
      keys: ['.', '.'],
      command: cmdIds.invokeNotebook
    },
```

Issue above, when I hit `..` on my keyboard, it will autocomplete the existing command, but it won't insert the text `.` in the line. Ex: `datetime.` is what I want, instead of `datetime`, `.` won't be inserted. 

I suspect that `.` become a special trigger instead of plain text. So I guess that what I want is hitting `.` and then trigger autocomplete.

I have not tested yet. Before I [invoke the autocomplete](https://github.com/mattma/jupyterlab/blob/master/examples/notebook/src/index.ts#L170), I should check the text value, if it is `.`, then insert as `dot`, then trigger autocomplete. Well, but I guess if i set it as `.` in the `keys: ['.'],` config, it is already as a special character.

**Comment by blink1073** [comment]
I agree that dot completion will need to be handled separately, reacting after the keyboard event has reached the editor instead of before.
